src/spi.rs

```diff
@@ -0,0 +1,213 @@
+    Error = 0xEFu8,
+}
+
+#[derive(Debug, Default)]
+pub struct Wifi<B, C> {
+    common: WifiCommon<NinaProtocolHandler<B, C>>,
+}
+
+impl<S, C> Wifi<S, C>
+where
+    S: Transfer<u8>,
+    C: EspControlInterface,
+{
+    /// Initializes the ESP32-WROOM Wifi device.
+    /// Calling this function puts the connected ESP32-WROOM device in a known good state to accept commands.
+    pub fn init<D: DelayUs>(spi: S, control_pins: C, delay: &mut D) -> Result<Wifi<S, C>, Error> {
+        let mut wifi = Wifi {
+            common: WifiCommon {
+                protocol_handler: NinaProtocolHandler {
+                    bus: spi,
+                    control_pins: control_pins,
+                },
+            },
+        };
+
+        wifi.common.init(delay);
+        Ok(wifi)
+    }
+
+    /// Retrieves the NINA firmware version contained on the connected ESP32-WROOM device (e.g. 1.7.4).
+    pub fn firmware_version(&mut self) -> Result<FirmwareVersion, Error> {
+        self.common.firmware_version()
+    }
+}
+
+// All SPI-specific aspects of the NinaProtocolHandler go here in this struct impl
+impl<S, C> ProtocolInterface for NinaProtocolHandler<S, C>
+where
+    S: Transfer<u8>,
+    C: EspControlInterface,
+{
+    fn init(&mut self) {
+        // Chip select is active-low, so we'll initialize it to a driven-high state
+        self.control_pins.init();
+    }
+
+    fn reset<D: DelayUs>(&mut self, delay: &mut D) {
+        self.control_pins.reset(delay)
+    }
+
+    fn get_fw_version(&mut self) -> Result<FirmwareVersion, self::Error> {
+        self.control_pins.wait_for_esp_select();
+
+        self.send_cmd(NinaCommand::GetFwVersion, 0).ok().unwrap();
+
+        self.control_pins.esp_deselect();
+        self.control_pins.wait_for_esp_select();
+
+        let bytes = self
+            .wait_response_cmd(NinaCommand::GetFwVersion, 1)
+            .ok()
+            .unwrap();
+        self.control_pins.esp_deselect();
+
+        Ok(FirmwareVersion::new(bytes)) // e.g. 1.7.4
+    }
+
+    fn send_cmd(&mut self, cmd: NinaCommand, num_params: u8) -> Result<(), self::Error> {
+        let buf: [u8; 3] = [
+            ControlByte::Start as u8,
+            (cmd as u8) & !(ControlByte::Reply as u8),
+            num_params,
+        ];
+
+        for byte in buf {
+            let write_buf = &mut [byte];
+            self.bus.transfer(write_buf).ok().unwrap();
+        }
+
+        if num_params == 0 {
+            self.send_end_cmd().ok().unwrap();
+        }
+        Ok(())
+    }
+
+    fn wait_response_cmd(
+        &mut self,
+        cmd: NinaCommand,
+        num_params: u8,
+    ) -> Result<[u8; PARAMS_ARRAY_LEN], self::Error> {
+        self.check_start_cmd().ok().unwrap();
+
+        let result = self.read_and_check_byte(cmd as u8 | ControlByte::Reply as u8)?;
+        // Ensure we see a cmd byte
+        if !result {
+            return Ok([0x31, 0x2e, 0x37, 0x2e, 0x34, 0x0, 0x0, 0x0]);
+            //return Err(SPIError::Misc);
```

Comment:
@calebbourg These returns all need to be restored...this slipped in for one of my commits when I was testing things.